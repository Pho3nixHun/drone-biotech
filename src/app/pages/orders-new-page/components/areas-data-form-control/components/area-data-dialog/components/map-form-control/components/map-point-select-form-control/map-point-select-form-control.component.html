@let vM = vm();
<div class="join" *transloco="let t">
    <app-button
        [vm]="vM.addButtonXVM"
        [disabled]="!!entryPointSignal()"
        (click)="drawMarker(null)"
    >
        @if (vM.addButtonXVM.textKey) {
            {{ t(vM.addButtonXVM.textKey) }}
        }
    </app-button>
    <app-button
        [vm]="vM.deleteButtonXVM"
        [disabled]="!entryPointSignal()"
        (click)="deleteMarker()"
    >
        @if (vM.deleteButtonXVM.textKey) {
            {{ t(vM.deleteButtonXVM.textKey) }}
        }
    </app-button>
</div>
