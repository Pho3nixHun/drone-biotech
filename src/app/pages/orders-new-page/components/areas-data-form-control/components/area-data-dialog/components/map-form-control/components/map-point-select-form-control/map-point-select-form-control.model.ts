import { InjectionToken } from '@angular/core';

export const ENTRY_POINT_MARKER_OPTIONS =
    new InjectionToken<google.maps.marker.AdvancedMarkerElementOptions>(
        'Injection Token for entry point marker options'
    );

export const provideMockEntryPointMarkerOptions = () => ({
    provide: ENTRY_POINT_MARKER_OPTIONS,
    useValue: {},
});

export interface MapPointSelectFormControlVM {
    deleteButtonTextKey: string;
    addButtonTextKey: string;
}
