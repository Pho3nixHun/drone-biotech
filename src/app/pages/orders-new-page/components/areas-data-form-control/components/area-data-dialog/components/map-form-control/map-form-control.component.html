<div class="flex flex-col space-y-2">
    <app-map-search-input-form-control
        [defaultCenter]="vm().defaultCenter"
        [vm]="vm().mapSearchInputFormControlVM"
        (changeBounds)="setMapBounds($event)"
    />

    <div appElementRef [signal]="divSignal" style="height: 600px"></div>

    <app-map-area-select-form-control
        [mapRef]="mapSignal()"
        [vm]="vm().mapAreaSelectFormControlVM"
        [formControl]="mapForm.controls.targetArea"
    />
    <app-map-point-select-form-control
        [mapRef]="mapSignal()"
        [vm]="vm().mapPointSelectFormControlVM"
        [formControl]="mapForm.controls.entryPoint"
    />
</div>
