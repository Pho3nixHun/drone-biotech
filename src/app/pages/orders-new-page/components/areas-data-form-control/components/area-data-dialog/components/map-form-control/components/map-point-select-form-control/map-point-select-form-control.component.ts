/* eslint-disable @typescript-eslint/no-empty-function */
import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
import { MapPointSelectFormControlService } from './map-point-select-form-control.service';
import { TranslocoModule } from '@jsverse/transloco';
import { MapPointSelectFormControlVM } from './map-point-select-form-control.model';
import { NgClass } from '@angular/common';
import { toSignal } from '@angular/core/rxjs-interop';
import { Coordinates } from '@stores/location/location.model';
import { Component, effect, forwardRef, inject, input } from '@angular/core';
import { ENTRY_POINT_MARKER_OPTIONS } from 'src/app/shared/providers/google-maps-provider';

const noop = () => {};

@Component({
    selector: 'app-map-point-select-form-control',
    imports: [TranslocoModule, NgClass],
    providers: [
        {
            provide: NG_VALUE_ACCESSOR,
            useExisting: forwardRef(() => MapPointSelectFormControlComponent),
            multi: true,
        },
        {
            provide: MapPointSelectFormControlService,
            deps: [ENTRY_POINT_MARKER_OPTIONS],
        },
    ],
    templateUrl: './map-point-select-form-control.component.html',
})
export class MapPointSelectFormControlComponent
    implements ControlValueAccessor
{
    private readonly pointSelectService = inject(
        MapPointSelectFormControlService
    );
    public readonly vm = input.required<MapPointSelectFormControlVM>();
    public readonly mapRef = input<google.maps.Map | null>();

    private readonly initializeMapEffect = effect(
        () => {
            const map = this.mapRef();
            if (!map) return;
            this.pointSelectService.initializeMap(map);
            this.initializeMapEffect.destroy();
        },
        { manualCleanup: true }
    );

    protected readonly entryPointSignal = toSignal(
        this.pointSelectService.entryPointMarkerPath$,
        { initialValue: null }
    );

    protected readonly entryPointEffect = effect(() => {
        this.onChange(this.entryPointSignal());
    });

    protected drawMarker(coordinates: Coordinates | null) {
        this.pointSelectService.drawMarker(coordinates);
    }
    protected deleteMarker() {
        this.pointSelectService.deleteMarker();
    }

    private onChange: (value: Coordinates | null) => void = noop;
    private onTouched: () => void = noop;

    public writeValue(value: Coordinates | null): void {
        if (value) return this.pointSelectService.drawMarker(value);
        this.pointSelectService.deleteMarker();
    }

    public registerOnChange(fn: (value: Coordinates | null) => void): void {
        this.onChange = fn;
    }

    public registerOnTouched(fn: () => void): void {
        this.onTouched = fn;
    }
}
