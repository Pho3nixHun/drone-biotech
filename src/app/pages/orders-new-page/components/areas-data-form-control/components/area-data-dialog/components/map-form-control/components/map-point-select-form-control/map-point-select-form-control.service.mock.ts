/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-empty-function */
import { Injectable, signal } from '@angular/core';
import { MapPointSelectFormControlService } from './map-point-select-form-control.service';
import { Observable } from 'rxjs';
import { toObservable } from '@angular/core/rxjs-interop';
import { Coordinates } from '@stores/location/location.model';

@Injectable({
    providedIn: 'root',
})
export class MapPointSelectFormControlMockService {
    public initializeMap(mapRef: google.maps.Map): void {}
    public entryPointMarkerPath$: Observable<Coordinates | null> =
        toObservable(entryPointSignal);
    public drawMarker(coordinates: Coordinates | null): void {}
    public deleteMarker(): void {}
    public getPath(
        pos:
            | google.maps.LatLng
            | google.maps.LatLngLiteral
            | google.maps.LatLngAltitudeLiteral
    ): Coordinates {
        return { lat: 10, lng: 10 };
    }
}

export const provideMapPointSelectFormControlMockService = () => ({
    provide: MapPointSelectFormControlService,
    useClass: MapPointSelectFormControlMockService,
});

const entryPointSignal = signal<Coordinates | null>(null);

export const updateEntryPointSignal = (entryPoint: Coordinates | null) =>
    entryPointSignal.set(entryPoint);
