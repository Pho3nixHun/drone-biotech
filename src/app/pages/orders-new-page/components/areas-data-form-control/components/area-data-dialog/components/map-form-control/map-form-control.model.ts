import { MapSearchInputFormControlVM } from 'src/app/pages/orders-new-page/components/areas-data-form-control/components/area-data-dialog/components/map-form-control/components/map-search-input-form-control/map-search-input-form-control.model';
import { MapAreaSelectFormControlVM } from './components/map-area-select-form-control/map-area-select-form-control.model';
import { MapPointSelectFormControlVM } from './components/map-point-select-form-control/map-point-select-form-control.model';
import { Coordinates } from '../../area-data-dialog.model';
import { InjectionToken } from '@angular/core';

export interface MapFormControlVM {
    defaultCenter: Coordinates | null;
    mapAreaSelectFormControlVM: MapAreaSelectFormControlVM;
    mapPointSelectFormControlVM: MapPointSelectFormControlVM;
    mapSearchInputFormControlVM: MapSearchInputFormControlVM;
}

export const MAP_OPTIONS = new InjectionToken<MapOptions>(
    'Injection Token for Map options'
);

export const provideMockMapOptions = () => ({
    provide: MAP_OPTIONS,
    useValue: {},
});
export interface MapOptions {
    mapTypeId: 'roadmap';
    disableDoubleClickZoom: boolean;
    clickableIcons: boolean;
    mapId: 'DEMO_MAP_ID';
    zoom: number;
    streetViewControl: boolean;
}

export const mapStringToCoordinates = (input: string): Coordinates[] => {
    const lines = input
        .split('\n')
        .map((line) => line.trim())
        .filter((line) => line !== ''); // ignore empty lines

    const coordinates: Coordinates[] = lines.map((line) => {
        const [lat, lng] = line.split(' ').map(Number);
        return { lat, lng };
    });

    return coordinates;
};

export const mapCoordinatesToString = (coordinates: Coordinates[]): string => {
    return coordinates.map((coord) => `${coord.lat} ${coord.lng}`).join('\n');
};
