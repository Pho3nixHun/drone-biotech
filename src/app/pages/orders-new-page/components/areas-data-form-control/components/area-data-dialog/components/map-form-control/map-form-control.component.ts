/* eslint-disable @typescript-eslint/no-empty-function */
import {
    Component,
    computed,
    effect,
    ElementRef,
    forwardRef,
    inject,
    input,
    Signal,
    signal,
} from '@angular/core';
import {
    ControlValueAccessor,
    FormBuilder,
    NG_VALUE_ACCESSOR,
    ReactiveFormsModule,
    Validators,
} from '@angular/forms';
import { MapFormControlVM } from './map-form-control.model';
import { Coordinates } from '@stores/location/location.model';
import { MapAreaSelectFormControlComponent } from './components/map-area-select-form-control/map-area-select-form-control.component';
import { MapPointSelectFormControlComponent } from './components/map-point-select-form-control/map-point-select-form-control.component';
import { MapSearchInputFormControlComponent } from './components/map-search-input-form-control/map-search-input-form-control.component';
import { distinctUntilChanged, map } from 'rxjs';
import { toSignal } from '@angular/core/rxjs-interop';
import { getCenter } from 'geolib';
import { ElementRefDirective } from '@directives/element-ref/element-ref.directive';
import {
    MAP_OPTIONS,
    HEAD_OFFICE_LOCATION,
} from '@providers/google-maps-provider';

@Component({
    selector: 'app-map-form-control',
    imports: [
        MapAreaSelectFormControlComponent,
        MapPointSelectFormControlComponent,
        ElementRefDirective,
        ReactiveFormsModule,
        MapSearchInputFormControlComponent,
    ],
    providers: [
        {
            provide: NG_VALUE_ACCESSOR,
            useExisting: forwardRef(() => MapFormControlComponent),
            multi: true,
        },
    ],
    schemas: [],
    templateUrl: './map-form-control.component.html',
})
export class MapFormControlComponent implements ControlValueAccessor {
    private readonly fb = inject(FormBuilder);
    private readonly mapOptions = inject(MAP_OPTIONS);
    protected readonly headOfficeLocation = inject(HEAD_OFFICE_LOCATION);
    public readonly vm = input.required<MapFormControlVM>();
    protected readonly divSignal = signal<ElementRef<HTMLElement> | null>(null);
    private readonly targetArea = signal<Coordinates[] | null>(null);

    protected readonly mapSignal = computed(() => {
        const canvas = this.divSignal();
        if (!canvas) return null;
        return new google.maps.Map(canvas.nativeElement);
    });

    private readonly targetAreaCenter = computed<Coordinates | null>(() => {
        const area = this.targetArea();
        if (!area) return null;
        const center = getCenter(area);
        return center
            ? {
                  lat: center.latitude,
                  lng: center.longitude,
              }
            : null;
    });

    protected readonly mapCenter = computed<Coordinates>(
        () =>
            this.targetAreaCenter() ??
            this.vm().defaultCenter ??
            this.headOfficeLocation
    );

    protected readonly setMapCenterEffect = effect(
        () => {
            const map = this.mapSignal();
            if (!map) return;
            map.setCenter(this.mapCenter());
            this.setMapCenterEffect.destroy();
        },
        { manualCleanup: true }
    );

    protected mapForm = this.fb.group({
        targetArea: this.fb.control<Coordinates[] | null>(
            null,
            Validators.required
        ),
        entryPoint: this.fb.control<Coordinates | null>(
            null,
            Validators.required
        ),
    });

    private readonly setMapOptionsEffect = effect(() => {
        const map = this.mapSignal();
        if (!map) return;

        map.setOptions({
            ...this.mapOptions,
            mapTypeControlOptions: {
                position: google.maps.ControlPosition.TOP_RIGHT,
            },
        });
    });

    protected setMapBounds(bounds: google.maps.LatLngBounds | null) {
        const map = this.mapSignal();
        if (!bounds || !map) return;
        map.fitBounds(bounds);
    }

    private readonly mapFormValue: Signal<{
        targetArea: Coordinates[] | null;
        entryPoint: Coordinates | null;
    }> = toSignal(
        this.mapForm.valueChanges.pipe(
            distinctUntilChanged(),
            map((data) => {
                const { entryPoint, targetArea } = data;
                return {
                    entryPoint: entryPoint ?? null,
                    targetArea: targetArea ?? null,
                };
            })
        ),
        { initialValue: { targetArea: null, entryPoint: null } }
    );

    private readonly mapFormValueEffect = effect(() =>
        this.onChange(this.mapFormValue())
    );

    private onChange: (value: {
        targetArea: Coordinates[] | null;
        entryPoint: Coordinates | null;
    }) => void = () => {};

    private onTouched: () => void = () => {};

    public writeValue(value: {
        targetArea: Coordinates[] | null;
        entryPoint: Coordinates | null;
    }) {
        this.mapForm.setValue(value);
        this.targetArea.set(value.targetArea);
    }

    public registerOnChange(fn: () => void): void {
        this.onChange = fn;
    }

    public registerOnTouched(fn: () => void): void {
        this.onTouched = fn;
    }
}
