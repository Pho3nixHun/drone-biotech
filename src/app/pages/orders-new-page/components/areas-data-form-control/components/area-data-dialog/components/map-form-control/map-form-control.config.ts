import { EnvironmentProviders, Provider } from '@angular/core';
import { HEAD_OFFICE_LOCATION } from '@services/distance/distance.model';
import {
    GOOGLE_MAPS_LOADER_CONFIG,
    provideGoogleMapsLibraries,
} from 'src/app/shared/providers/google-maps-provider';
import { environment } from 'src/environments/environment';
import {
    POLYGON_OPTIONS,
    INFO_WINDOW_OPTIONS,
} from './components/map-area-select-form-control/map-area-select-form-control.model';
import { ENTRY_POINT_MARKER_OPTIONS } from './components/map-point-select-form-control/map-point-select-form-control.model';
import { MAP_OPTIONS } from './map-form-control.model';

export const mapFormControlConfig: (Provider | EnvironmentProviders)[] = [
    provideGoogleMapsLibraries(),
    {
        provide: POLYGON_OPTIONS,
        useValue: {
            clickable: false,
            draggable: false,
            editable: true,
            fillOpacity: 0.5,
            strokeColor: 'blue',
            fillColor: 'blue',
        },
    },
    {
        provide: MAP_OPTIONS,
        useValue: {
            mapTypeId: 'roadmap',
            disableDoubleClickZoom: true,
            clickableIcons: false,
            mapId: 'DEMO_MAP_ID',
            zoom: 14,
            streetViewControl: false,
        },
    },
    {
        provide: ENTRY_POINT_MARKER_OPTIONS,
        useValue: {
            gmpDraggable: true,
        },
    },
    {
        provide: GOOGLE_MAPS_LOADER_CONFIG,
        useValue: {
            url: 'https://maps.googleapis.com/maps/api/js?',
            key: environment.googleMapsConfig.apiKey,
            version: '3.59',
            region: 'HU',
        },
    },
    {
        provide: HEAD_OFFICE_LOCATION,
        useValue: { lat: 47.312498121576795, lng: 21.309304570654604 },
    },
    {
        provide: INFO_WINDOW_OPTIONS,
        useValue: {
            disableAutoPan: true,
            headerDisabled: true,
        },
    },
];
