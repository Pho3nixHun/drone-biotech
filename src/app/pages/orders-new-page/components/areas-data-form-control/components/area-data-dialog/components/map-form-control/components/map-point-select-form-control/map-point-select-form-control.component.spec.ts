import { ComponentFixture, TestBed } from '@angular/core/testing';
import { MapPointSelectFormControlComponent } from './map-point-select-form-control.component';
import {
    MapPointSelectFormControlVM,
    provideMockEntryPointMarkerOptions,
} from './map-point-select-form-control.model';
import {
    Component,
    computed,
    DebugElement,
    ElementRef,
    inject,
    input,
    signal,
} from '@angular/core';
import { By } from '@angular/platform-browser';
import { MapPointSelectFormControlService } from './map-point-select-form-control.service';
import {
    MapPointSelectFormControlMockService,
    provideMapPointSelectFormControlMockService,
} from './map-point-select-form-control.service.mock';
import { ElementRefDirective } from '@directives/element-ref/element-ref.directive';
import { provideMockHeadOfficeLocation } from '@services/distance/distance.model';
import { provideMockMapOptions } from '../../map-form-control.model';
import { FormBuilder, ReactiveFormsModule } from '@angular/forms';
import { Coordinates } from '@stores/location/location.model';
import { MatIcon } from '@interfaces/mat-icon.enum';
import { getTranslocoModule } from 'transloco-testing.module';

const vm: MapPointSelectFormControlVM = {
    addButtonXVM: { icon: MatIcon.ADD, variant: 'fill' },
    deleteButtonXVM: { icon: MatIcon.ADD, variant: 'fill' },
};

@Component({
    imports: [
        MapPointSelectFormControlComponent,
        ElementRefDirective,
        ReactiveFormsModule,
    ],
    template: `
        <div appElementRef [signal]="divSignal" style="height: 600px"></div>

        <app-map-point-select-form-control
            [mapRef]="mapSignal()"
            [vm]="vm()"
            [formControl]="control"
        />
    `,
})
class TestHostComponent {
    private readonly fb = inject(FormBuilder);
    public vm = input.required<MapPointSelectFormControlVM>();
    public control = this.fb.control<Coordinates | null>(null);
    protected readonly divSignal = signal<ElementRef<HTMLElement> | null>(null);
    protected readonly mapSignal = computed(() => {
        const canvas = this.divSignal();
        if (!canvas) return null;
        return new google.maps.Map(canvas.nativeElement);
    });
}

describe('MapPointSelectFormControlComponent', () => {
    let fixture: ComponentFixture<TestHostComponent>;
    let compiled: HTMLElement;
    let mapPointSelectFormControlComponent: MapPointSelectFormControlComponent;
    let innerDebugElement: DebugElement;

    beforeEach(async () => {
        await TestBed.configureTestingModule({
            imports: [
                TestHostComponent,
                getTranslocoModule({
                    langs: { en: {} },
                    translocoConfig: {
                        availableLangs: ['en'],
                        defaultLang: 'en',
                    },
                }),
            ],
            providers: [
                provideMockEntryPointMarkerOptions(),
                provideMockMapOptions(),
                provideMockHeadOfficeLocation(),
                provideMapPointSelectFormControlMockService(),
            ],
        }).compileComponents();
        TestBed.overrideProvider(MapPointSelectFormControlService, {
            useFactory: () => new MapPointSelectFormControlMockService(),
        });
        fixture = TestBed.createComponent(TestHostComponent);
        fixture.componentRef.setInput('vm', vm);
        innerDebugElement = fixture.debugElement.query(
            By.directive(MapPointSelectFormControlComponent)
        );
        mapPointSelectFormControlComponent =
            innerDebugElement.componentInstance;
        compiled = innerDebugElement.nativeElement;
    });

    // Snapshot test
    it('should render the template when the VM is provided', () => {
        //Arrange

        //Act
        fixture.detectChanges();

        //Assert
        expect(compiled).toMatchSnapshot();
    });

    // Unit test
    it('should initialize the map', () => {
        //Arrange
        const serviceSpy = jest.spyOn(
            mapPointSelectFormControlComponent['pointSelectService'],
            'initializeMap'
        );

        //Act
        fixture.detectChanges();

        // Assert
        expect(serviceSpy).toHaveBeenCalledWith(
            mapPointSelectFormControlComponent['mapRef']()
        );
    });
});
