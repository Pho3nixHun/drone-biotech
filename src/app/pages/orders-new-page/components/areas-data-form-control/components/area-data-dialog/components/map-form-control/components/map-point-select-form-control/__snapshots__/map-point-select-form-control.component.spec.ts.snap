// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapPointSelectFormControlComponent should render the template when the VM is provided 1`] = `
<div
  id="root0"
  ng-version="18.2.8"
>
  <app-map-point-select-form-control
    ng-reflect-vm="[object Object]"
  >
    <div
      class="flex flex-col space-y-4 bg-white"
    >
      <div
        class="w-full h-[40vh]"
        ng-reflect-attached-callback="async (mapCanvas) => {
       "
      />
      <button
        class="btn btn-primary"
        ng-reflect-ng-class="btn-primary"
      >
         addButton 
      </button>
      <button
        class="btn btn-disabled"
        disabled=""
        ng-reflect-ng-class="btn-disabled"
      >
         deleteButton 
      </button>
    </div>
    <!--container-->
  </app-map-point-select-form-control>
</div>
`;
